pallets/rc-migrator/src/staking/nom_pools.rs

```diff
@@ -328,3 +334,284 @@ impl<T: pallet_nomination_pools::Config> NomPoolsMigrator<T> {
+	AccountId: Debug + Clone + PartialEq,
+	BlockNumber: Debug + Clone + PartialEq,
+> {
+	/// All `StorageValues` that can be migrated at once.
+	StorageValues { values: NomPoolsStorageValues<Balance> },
+	/// Entry of the `PoolMembers` map.
+	PoolMembers { member: (AccountId, GenericPoolMember<Balance, RewardCounter>) },
+	/// Entry of the `BondedPools` map.
+	BondedPools { pool: (PoolId, GenericBondedPoolInner<Balance, AccountId, BlockNumber>) },
+	/// Entry of the `RewardPools` map.
+	RewardPools { rewards: (PoolId, GenericRewardPool<Balance, RewardCounter>) },
+	/// Entry of the `SubPoolsStorage` map.
+	SubPoolsStorage { sub_pools: (PoolId, GenericSubPools<Balance>) },
+	/// Entry of the `Metadata` map.
+	Metadata { meta: (PoolId, BoundedVec<u8, ConstU32<256>>) },
+	/// Entry of the `ReversePoolIdLookup` map.
+	// TODO check if inserting None into an option map is the same as deleting the key
+	ReversePoolIdLookup { lookups: (AccountId, PoolId) },
+	/// Entry of the `ClaimPermissions` map.
+	ClaimPermissions { perms: (AccountId, ClaimPermission) },
+}
+
+#[cfg(feature = "std")]
+impl<T: Config> crate::types::RcMigrationCheck for NomPoolsMigrator<T> {
+	type RcPrePayload = Vec<
+		GenericNomPoolsMessage<
+			BalanceOf<T>,
+			T::RewardCounter,
+			<T as frame_system::Config>::AccountId,
+			BlockNumberFor<T>,
+		>,
+	>;
+
+	fn pre_check() -> Self::RcPrePayload {
+		let mut messages = Vec::new();
+
+		// Collect storage values
+		let values = NomPoolsStorageValues {
+			total_value_locked: pallet_nomination_pools::TotalValueLocked::<T>::get(),
+			min_join_bond: pallet_nomination_pools::MinJoinBond::<T>::get(),
+			min_create_bond: pallet_nomination_pools::MinCreateBond::<T>::get(),
+			max_pools: pallet_nomination_pools::MaxPools::<T>::get(),
+			max_pool_members: pallet_nomination_pools::MaxPoolMembers::<T>::get(),
+			max_pool_members_per_pool: pallet_nomination_pools::MaxPoolMembersPerPool::<T>::get(),
+			global_max_commission: pallet_nomination_pools::GlobalMaxCommission::<T>::get(),
+			last_pool_id: pallet_nomination_pools::LastPoolId::<T>::get(),
+		};
+		messages.push(GenericNomPoolsMessage::StorageValues { values });
+
+		// Collect pool members
+		for (who, member) in pallet_nomination_pools::PoolMembers::<T>::iter() {
+			let generic_member = GenericPoolMember {
+				pool_id: member.pool_id,
+				points: member.points,
+				last_recorded_reward_counter: member.last_recorded_reward_counter,
+				unbonding_eras: member.unbonding_eras.into_inner(),
+			};
+			messages.push(GenericNomPoolsMessage::PoolMembers { member: (who, generic_member) });
+		}
+
+		// Collect bonded pools
+		for (pool_id, pool) in pallet_nomination_pools::BondedPools::<T>::iter() {
+			let generic_pool = GenericBondedPoolInner {
+				commission: GenericCommission {
+					current: pool.commission.current,
+					max: pool.commission.max,
+					change_rate: pool.commission.change_rate,
+					throttle_from: None, // None to avoid discrepancies during the AH postcheck
```

Comment:
I would calculate here what is expected on post check and not follow what the migration does (extract with RC pallet, map with AH pallet). Like with AAA tests, pre_check here for me would be first A, second A would be the migration itself, and last A would be in AH post check.

Reply:
You're right. I moved the calculation of `change_rate` to the RC pre-check and removed `normalize_generic_bonded_pool` function from AH